Mathlib/Topology/MetricSpace/Ultra/Basic.lean

```diff
@@ -0,0 +1,234 @@
+/-
+Copyright (c) 2024 Yakov Pechersky. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Yakov Pechersky
+-/
+import Mathlib.Topology.MetricSpace.Pseudo.Defs
+
+/-!
+## Ultrametric spaces
+
+This file defines ultrametric spaces, implemented as a mixin on the `Dist`,
+so that it can apply on pseudometric spaces as well.
+
+## Main definitions
+
+* `IsUltrametricDist X`: Annotates `dist : X → X → ℝ` as respecting the ultrametric inequality
+  of `dist(x, z) ≤ max {dist(x,y), dist(y,z)}`
+
+## Implementation details
+
+The mixin could have been defined as a hypothesis to be carried around, instead of relying on
+typeclass synthesis. However, since we declare a (pseudo)metric space on a type using
+typeclass arguments, one can declare the ultrametricity at the same time.
```

Comment:
Not really clear what you mean; can you expand a bit?

Reply:
I looked at how nonarchimedean norms were specified elsewhere, and I saw it done using an explicit IsNonarchimedean hypothesis here: https://github.com/leanprover-community/mathlib4/blob/master/Mathlib%2FTopology%2FAlgebra%2FValued%2FNormedValued.lean#L32

There is also the option to work over Valued spaces instead of modifying the dist or the norm.

Reply:
Added a comment